### ISSUE TITLE
route transform performance regression in 0.20.0

### ISSUE THREAD

**Issue by debugmiller** [vectordotdev/vector#11997]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

We have a `route` transform with 50 different routes. We run this in vector with 8 threads (`-t8`). Prior to 0.20.0 we could maximize CPU fully (800%), but after 0.20.0 we cannot get above ~170%. I suspect this to be related to the changes to `route` which turn it into a single transform internally. We can obviously run with 1 thread to get similar performance, but our deployment (kubernetes) has per replica overhead which is better to spread out across instances that are multithreaded. Is it expected for this transform to scale gracefully across many routes and many cores?

### Version

vector 0.20.0 (x86_64-unknown-linux-gnu 2a706a3 2022-02-11)

**Comment by jszwedko** [comment]
Hi @debugmiller !

That definitely wasn't expected. Could you share an example config that demonstrates the regression you are seeing? That should help us reproduce and narrow it down. I'm wondering if the configuration is bottlenecked in a different place that is causing `route` to be under utilized.

You could also compare the `utilization` metric before and after to see which components are the bottleneck (they will be last components in the topology that have a utilization close to `1`).

**Comment by jszwedko** [comment]
As you noted, the transform was reworked to run as a single transform in 0.20.0 rather than expanding out to N transforms (one per route), but the transform itself will still run in concurrently, and in parallel.

**Comment by debugmiller** [comment]
My toml is about 2000 lines long but heres the utilization from one of the instances the blue line is the `route` transform
![image](https://user-images.githubusercontent.com/40829765/160218802-e390df18-0ac3-424d-b1f8-81241e031857.png)

**Comment by debugmiller** [comment]
My config is 1 kafka source that sends all data to 1 route. That route has 50 branches. Each branch goes to its own remap transform, the output of these 50 transforms is then collected by a single remap, that then sends them all to a splunk sink.

**Comment by debugmiller** [comment]
I tried to produce an example with `demo_logs` -> `route` -> `blackhole` but I couldn't get it to behave similarly

**Comment by debugmiller** [comment]
Here's also the CPU usage graph for the same timeframe as the above graph. It shows the 0.19.0, failed 0.20.0 deploy, and then the rollback. All with the same config. Prior to 15:30 you can see the 0.19.0 deploy, then the rollout to 0.20.0 just before 15:30 and then the rollback at 15:45
![image](https://user-images.githubusercontent.com/40829765/160220224-5c181227-1ef7-4646-ac01-78a37df99dc2.png)

**Comment by debugmiller** [comment]
Here is the route transform with the only changes I made from our live deploy are the names of routes and string literals. There is a silly hack for a "null" route that I left in for completeness but I don't think it's relevant.
```
[transforms.topic]
    type = "route"
    inputs = [ "*-kafka-src" ]

    [transforms.topic.route]
    alpha = '.topic == "alpha"'
    bravo = '.topic == "bravo"'
    charlie = '.topic == "charlie"'
    delta = '.topic == "delta"'
    ...[many omitted]...
    zulu = '.topic == "zulu"'
    
    # this is a null route, helpful for preventing 'transform has no input' errors
    null = '.topic == "thisismadeup"'
```

**Comment by hhromic** [comment]
Hello, I made my own benchmarking about your issue because we are heavy users of the `route` transform aswell.
I setup an exaggerated MRE (using 100 routes) and I can also fully confirm your observations.
I also compared the `route` transform using the `check_fields` condition type instead of `vrl` and despite obtaining better performance I also could observe a difference between `0.20.0` and `0.19.3`. See below.

In summary, I can observe the following average throughput difference between versions:
* `~47.1%` drop for `vrl` type conditions.
* `~16.6%` drop for `check_fields` type conditions.

### Configuration Templates

To generate the following MRE pipelines, I used the [`ytt`](https://github.com/vmware-tanzu/carvel-ytt) templating tool. For example:
```
$ ytt -f vector.tmpl.yaml > vector.yaml
$ ytt -f vector-chkfld.tmpl.yaml > vector-chkfld.yaml
```

The `ytt` YAML template for a `vrl` condition type `route`  is:
```yaml
#@ NUM_ROUTES = 100
#@yaml/text-templated-strings
---
api:
  enabled: true
  address: 0.0.0.0:8686
  playground: false

sources:
  logs:
    type: demo_logs
    format: json
    interval: 0

transforms:
  add-key:
    type: remap
    inputs:
      - logs
    source: |-
      .key = to_unix_timestamp(now(), unit: "milliseconds") % (@= str(NUM_ROUTES) @)
  route:
    type: route
    inputs:
      - add-key
    route:
      #@ for idx in range(NUM_ROUTES):
      route(@= str(idx) @): '.key == (@= str(idx) @)'
      #@ end

sinks:
  blackhole:
    type: blackhole
    inputs:
      - 'route.route*'
```

And the `ytt` YAML template for a `check_fields` condition type `route`  is:
```yaml
#@ NUM_ROUTES = 100
#@yaml/text-templated-strings
---
api:
  enabled: true
  address: 0.0.0.0:8686
  playground: false

sources:
  logs:
    type: demo_logs
    format: json
    interval: 0

transforms:
  add-key:
    type: remap
    inputs:
      - logs
    source: |-
      .key = to_unix_timestamp(now(), unit: "milliseconds") % (@= str(NUM_ROUTES) @)
  route:
    type: route
    inputs:
      - add-key
    route:
      #@ for idx in range(NUM_ROUTES):
      route(@= str(idx) @):
        type: check_fields
        key.eq: #@ idx
      #@ end

sinks:
  blackhole:
    type: blackhole
    inputs:
      - 'route.route*'
```

### Results using Vector `0.19.3`

For the `vrl` condition type:
```
2022-03-30T09:52:36.461174Z  INFO vector::sinks::blackhole::sink: Total events collected events=0 raw_bytes_collected=0
2022-03-30T09:52:37.461350Z  INFO vector::sinks::blackhole::sink: Total events collected events=6769 raw_bytes_collected=5582741
2022-03-30T09:52:38.462719Z  INFO vector::sinks::blackhole::sink: Total events collected events=12259 raw_bytes_collected=10110077
2022-03-30T09:52:39.461379Z  INFO vector::sinks::blackhole::sink: Total events collected events=17628 raw_bytes_collected=14536881
2022-03-30T09:52:40.462125Z  INFO vector::sinks::blackhole::sink: Total events collected events=22822 raw_bytes_collected=18819122
2022-03-30T09:52:41.462684Z  INFO vector::sinks::blackhole::sink: Total events collected events=28166 raw_bytes_collected=23225066
2022-03-30T09:52:42.461670Z  INFO vector::sinks::blackhole::sink: Total events collected events=33486 raw_bytes_collected=27611855
2022-03-30T09:52:43.465953Z  INFO vector::sinks::blackhole::sink: Total events collected events=38963 raw_bytes_collected=32125388
2022-03-30T09:52:44.460219Z  INFO vector::sinks::blackhole::sink: Total events collected events=43408 raw_bytes_collected=35790425
2022-03-30T09:52:45.464780Z  INFO vector::sinks::blackhole::sink: Total events collected events=48544 raw_bytes_collected=40024013
2022-03-30T09:52:46.461694Z  INFO vector::sinks::blackhole::sink: Total events collected events=52573 raw_bytes_collected=43344934
2022-03-30T09:52:47.460712Z  INFO vector::sinks::blackhole::sink: Total events collected events=55297 raw_bytes_collected=45589800
2022-03-30T09:52:48.460962Z  INFO vector::sinks::blackhole::sink: Total events collected events=58473 raw_bytes_collected=48206087
2022-03-30T09:52:49.469690Z  INFO vector::sinks::blackhole::sink: Total events collected events=61496 raw_bytes_collected=50697742
2022-03-30T09:52:50.460453Z  INFO vector::sinks::blackhole::sink: Total events collected events=66664 raw_bytes_collected=54958365
2022-03-30T09:52:51.464487Z  INFO vector::sinks::blackhole::sink: Total events collected events=71794 raw_bytes_collected=59188390
2022-03-30T09:52:52.461195Z  INFO vector::sinks::blackhole::sink: Total events collected events=76750 raw_bytes_collected=63273194
2022-03-30T09:52:53.468093Z  INFO vector::sinks::blackhole::sink: Total events collected events=82302 raw_bytes_collected=67851659
2022-03-30T09:52:54.461220Z  INFO vector::sinks::blackhole::sink: Total events collected events=87614 raw_bytes_collected=72232238
2022-03-30T09:52:55.461218Z  INFO vector::sinks::blackhole::sink: Total events collected events=93096 raw_bytes_collected=76750969
2022-03-30T09:52:56.460568Z  INFO vector::sinks::blackhole::sink: Total events collected events=98435 raw_bytes_collected=81154713
2022-03-30T09:52:57.463816Z  INFO vector::sinks::blackhole::sink: Total events collected events=103821 raw_bytes_collected=85594797
```

For the `check_fields` condition type:
```
2022-03-30T10:00:01.658198Z  INFO vector::sinks::blackhole::sink: Total events collected events=0 raw_bytes_collected=0
2022-03-30T10:00:02.651827Z  INFO vector::sinks::blackhole::sink: Total events collected events=10280 raw_bytes_collected=8502047
2022-03-30T10:00:03.651952Z  INFO vector::sinks::blackhole::sink: Total events collected events=18686 raw_bytes_collected=15451225
2022-03-30T10:00:04.651790Z  INFO vector::sinks::blackhole::sink: Total events collected events=25225 raw_bytes_collected=20856357
2022-03-30T10:00:05.651997Z  INFO vector::sinks::blackhole::sink: Total events collected events=31925 raw_bytes_collected=26393387
2022-03-30T10:00:06.652291Z  INFO vector::sinks::blackhole::sink: Total events collected events=38890 raw_bytes_collected=32153915
2022-03-30T10:00:07.652025Z  INFO vector::sinks::blackhole::sink: Total events collected events=45922 raw_bytes_collected=37965194
2022-03-30T10:00:08.652517Z  INFO vector::sinks::blackhole::sink: Total events collected events=52964 raw_bytes_collected=43788458
2022-03-30T10:00:09.654211Z  INFO vector::sinks::blackhole::sink: Total events collected events=59607 raw_bytes_collected=49282243
2022-03-30T10:00:10.651996Z  INFO vector::sinks::blackhole::sink: Total events collected events=66538 raw_bytes_collected=55012175
2022-03-30T10:00:11.651768Z  INFO vector::sinks::blackhole::sink: Total events collected events=73423 raw_bytes_collected=60700659
2022-03-30T10:00:12.652471Z  INFO vector::sinks::blackhole::sink: Total events collected events=80646 raw_bytes_collected=66673375
2022-03-30T10:00:13.651854Z  INFO vector::sinks::blackhole::sink: Total events collected events=87755 raw_bytes_collected=72550520
2022-03-30T10:00:14.652337Z  INFO vector::sinks::blackhole::sink: Total events collected events=94482 raw_bytes_collected=78110869
2022-03-30T10:00:15.651708Z  INFO vector::sinks::blackhole::sink: Total events collected events=101295 raw_bytes_collected=83743165
```

### Results using Vector `0.20.0`

For the `vrl` condition type:
```
2022-03-30T09:51:59.878265Z  INFO vector::sinks::blackhole::sink: Total events collected events=0 raw_bytes_collected=0
2022-03-30T09:52:00.878763Z  INFO vector::sinks::blackhole::sink: Total events collected events=3578 raw_bytes_collected=2972638
2022-03-30T09:52:01.878978Z  INFO vector::sinks::blackhole::sink: Total events collected events=7365 raw_bytes_collected=6118136
2022-03-30T09:52:02.878439Z  INFO vector::sinks::blackhole::sink: Total events collected events=10651 raw_bytes_collected=8848204
2022-03-30T09:52:03.879310Z  INFO vector::sinks::blackhole::sink: Total events collected events=11879 raw_bytes_collected=9868566
2022-03-30T09:52:04.878976Z  INFO vector::sinks::blackhole::sink: Total events collected events=14902 raw_bytes_collected=12379568
2022-03-30T09:52:05.879164Z  INFO vector::sinks::blackhole::sink: Total events collected events=18318 raw_bytes_collected=15217778
2022-03-30T09:52:06.878397Z  INFO vector::sinks::blackhole::sink: Total events collected events=21886 raw_bytes_collected=18181576
2022-03-30T09:52:07.878971Z  INFO vector::sinks::blackhole::sink: Total events collected events=25469 raw_bytes_collected=21157852
2022-03-30T09:52:08.878891Z  INFO vector::sinks::blackhole::sink: Total events collected events=29172 raw_bytes_collected=24234666
2022-03-30T09:52:09.878071Z  INFO vector::sinks::blackhole::sink: Total events collected events=32795 raw_bytes_collected=27243218
2022-03-30T09:52:10.878613Z  INFO vector::sinks::blackhole::sink: Total events collected events=36267 raw_bytes_collected=30125983
2022-03-30T09:52:11.878113Z  INFO vector::sinks::blackhole::sink: Total events collected events=39596 raw_bytes_collected=32891296
2022-03-30T09:52:12.878458Z  INFO vector::sinks::blackhole::sink: Total events collected events=42887 raw_bytes_collected=35624706
2022-03-30T09:52:13.879039Z  INFO vector::sinks::blackhole::sink: Total events collected events=46039 raw_bytes_collected=38242745
2022-03-30T09:52:14.878411Z  INFO vector::sinks::blackhole::sink: Total events collected events=49466 raw_bytes_collected=41090002
2022-03-30T09:52:15.878738Z  INFO vector::sinks::blackhole::sink: Total events collected events=52884 raw_bytes_collected=43929924
2022-03-30T09:52:16.879042Z  INFO vector::sinks::blackhole::sink: Total events collected events=56254 raw_bytes_collected=46729513
2022-03-30T09:52:17.878078Z  INFO vector::sinks::blackhole::sink: Total events collected events=59651 raw_bytes_collected=49551639
2022-03-30T09:52:18.879039Z  INFO vector::sinks::blackhole::sink: Total events collected events=63096 raw_bytes_collected=52414560
2022-03-30T09:52:19.878333Z  INFO vector::sinks::blackhole::sink: Total events collected events=66372 raw_bytes_collected=55136556
2022-03-30T09:52:20.878351Z  INFO vector::sinks::blackhole::sink: Total events collected events=69848 raw_bytes_collected=58025047
2022-03-30T09:52:21.878819Z  INFO vector::sinks::blackhole::sink: Total events collected events=73254 raw_bytes_collected=60854611
2022-03-30T09:52:22.878278Z  INFO vector::sinks::blackhole::sink: Total events collected events=76614 raw_bytes_collected=63645525
2022-03-30T09:52:23.878487Z  INFO vector::sinks::blackhole::sink: Total events collected events=79980 raw_bytes_collected=66442421
2022-03-30T09:52:24.878489Z  INFO vector::sinks::blackhole::sink: Total events collected events=83006 raw_bytes_collected=68956981
2022-03-30T09:52:25.878753Z  INFO vector::sinks::blackhole::sink: Total events collected events=85683 raw_bytes_collected=71181093
2022-03-30T09:52:26.878492Z  INFO vector::sinks::blackhole::sink: Total events collected events=89108 raw_bytes_collected=74026052
2022-03-30T09:52:27.878288Z  INFO vector::sinks::blackhole::sink: Total events collected events=92505 raw_bytes_collected=76848160
2022-03-30T09:52:28.878158Z  INFO vector::sinks::blackhole::sink: Total events collected events=95935 raw_bytes_collected=79697590
2022-03-30T09:52:29.878830Z  INFO vector::sinks::blackhole::sink: Total events collected events=99440 raw_bytes_collected=82610150
2022-03-30T09:52:30.878336Z  INFO vector::sinks::blackhole::sink: Total events collected events=102652 raw_bytes_collected=85278656
```

For the `check_fields` condition type:
```
2022-03-30T09:59:30.385572Z  INFO vector::sinks::blackhole::sink: Total events collected events=0 raw_bytes_collected=0
2022-03-30T09:59:31.386171Z  INFO vector::sinks::blackhole::sink: Total events collected events=8567 raw_bytes_collected=7117014
2022-03-30T09:59:32.386082Z  INFO vector::sinks::blackhole::sink: Total events collected events=17592 raw_bytes_collected=14613327
2022-03-30T09:59:33.385151Z  INFO vector::sinks::blackhole::sink: Total events collected events=26663 raw_bytes_collected=22148820
2022-03-30T09:59:34.385343Z  INFO vector::sinks::blackhole::sink: Total events collected events=35891 raw_bytes_collected=29814543
2022-03-30T09:59:35.386062Z  INFO vector::sinks::blackhole::sink: Total events collected events=45009 raw_bytes_collected=37388716
2022-03-30T09:59:36.385097Z  INFO vector::sinks::blackhole::sink: Total events collected events=54104 raw_bytes_collected=44942416
2022-03-30T09:59:37.386071Z  INFO vector::sinks::blackhole::sink: Total events collected events=63090 raw_bytes_collected=52406546
2022-03-30T09:59:38.385865Z  INFO vector::sinks::blackhole::sink: Total events collected events=72166 raw_bytes_collected=59944896
2022-03-30T09:59:39.385268Z  INFO vector::sinks::blackhole::sink: Total events collected events=81324 raw_bytes_collected=67552814
2022-03-30T09:59:40.385309Z  INFO vector::sinks::blackhole::sink: Total events collected events=89782 raw_bytes_collected=74579421
2022-03-30T09:59:41.385227Z  INFO vector::sinks::blackhole::sink: Total events collected events=99126 raw_bytes_collected=82338761
2022-03-30T09:59:42.385286Z  INFO vector::sinks::blackhole::sink: Total events collected events=107307 raw_bytes_collected=89136056
```

**Comment by debugmiller** [comment]
@jszwedko is there any other information I can provide to help with investigating this issue?

**Comment by jszwedko** [comment]
Hey! No, this is helpful, especially @hhromic 's reproduction case. This should give us something to dig into.

**Comment by hhromic** [comment]
Hi again. I now re-ran the same MREs I reported before using the newly released `0.20.1` version.

For the `vrl` condition type:
```
2022-04-11T10:14:07.253962Z  INFO vector::sinks::blackhole::sink: Total events collected events=0 raw_bytes_collected=0
2022-04-11T10:14:08.254080Z  INFO vector::sinks::blackhole::sink: Total events collected events=4286 raw_bytes_collected=3561229
2022-04-11T10:14:09.253443Z  INFO vector::sinks::blackhole::sink: Total events collected events=8730 raw_bytes_collected=7252530
2022-04-11T10:14:10.254248Z  INFO vector::sinks::blackhole::sink: Total events collected events=13266 raw_bytes_collected=11021028
2022-04-11T10:14:11.254221Z  INFO vector::sinks::blackhole::sink: Total events collected events=17909 raw_bytes_collected=14878863
2022-04-11T10:14:12.254123Z  INFO vector::sinks::blackhole::sink: Total events collected events=22379 raw_bytes_collected=18590789
2022-04-11T10:14:13.253872Z  INFO vector::sinks::blackhole::sink: Total events collected events=26705 raw_bytes_collected=22185589
2022-04-11T10:14:14.253738Z  INFO vector::sinks::blackhole::sink: Total events collected events=30949 raw_bytes_collected=25712723
2022-04-11T10:14:15.253466Z  INFO vector::sinks::blackhole::sink: Total events collected events=35276 raw_bytes_collected=29307665
2022-04-11T10:14:16.254141Z  INFO vector::sinks::blackhole::sink: Total events collected events=39569 raw_bytes_collected=32874141
2022-04-11T10:14:17.253570Z  INFO vector::sinks::blackhole::sink: Total events collected events=43849 raw_bytes_collected=36430071
2022-04-11T10:14:18.253341Z  INFO vector::sinks::blackhole::sink: Total events collected events=48155 raw_bytes_collected=40007872
2022-04-11T10:14:19.253505Z  INFO vector::sinks::blackhole::sink: Total events collected events=51974 raw_bytes_collected=43180857
2022-04-11T10:14:20.253864Z  INFO vector::sinks::blackhole::sink: Total events collected events=56075 raw_bytes_collected=46587126
2022-04-11T10:14:21.253467Z  INFO vector::sinks::blackhole::sink: Total events collected events=60309 raw_bytes_collected=50105034
2022-04-11T10:14:22.253473Z  INFO vector::sinks::blackhole::sink: Total events collected events=64571 raw_bytes_collected=53645216
2022-04-11T10:14:23.253306Z  INFO vector::sinks::blackhole::sink: Total events collected events=68839 raw_bytes_collected=57191228
2022-04-11T10:14:24.253264Z  INFO vector::sinks::blackhole::sink: Total events collected events=73107 raw_bytes_collected=60737205
2022-04-11T10:14:25.253817Z  INFO vector::sinks::blackhole::sink: Total events collected events=77446 raw_bytes_collected=64342142
2022-04-11T10:14:26.253599Z  INFO vector::sinks::blackhole::sink: Total events collected events=81657 raw_bytes_collected=67839740
2022-04-11T10:14:27.253921Z  INFO vector::sinks::blackhole::sink: Total events collected events=85963 raw_bytes_collected=71416866
2022-04-11T10:14:28.253573Z  INFO vector::sinks::blackhole::sink: Total events collected events=90163 raw_bytes_collected=74906479
2022-04-11T10:14:29.253774Z  INFO vector::sinks::blackhole::sink: Total events collected events=93700 raw_bytes_collected=77845265
2022-04-11T10:14:30.253644Z  INFO vector::sinks::blackhole::sink: Total events collected events=97419 raw_bytes_collected=80934160
2022-04-11T10:14:31.253646Z  INFO vector::sinks::blackhole::sink: Total events collected events=101472 raw_bytes_collected=84300840
```

For the `check_fields` condition type:
```
2022-04-11T10:15:22.768941Z  INFO vector::sinks::blackhole::sink: Total events collected events=0 raw_bytes_collected=0
2022-04-11T10:15:23.769278Z  INFO vector::sinks::blackhole::sink: Total events collected events=11281 raw_bytes_collected=9370257
2022-04-11T10:15:24.769489Z  INFO vector::sinks::blackhole::sink: Total events collected events=22601 raw_bytes_collected=18772979
2022-04-11T10:15:25.769377Z  INFO vector::sinks::blackhole::sink: Total events collected events=34095 raw_bytes_collected=28319990
2022-04-11T10:15:26.768699Z  INFO vector::sinks::blackhole::sink: Total events collected events=45393 raw_bytes_collected=37705994
2022-04-11T10:15:27.769316Z  INFO vector::sinks::blackhole::sink: Total events collected events=56794 raw_bytes_collected=47177582
2022-04-11T10:15:28.768996Z  INFO vector::sinks::blackhole::sink: Total events collected events=67858 raw_bytes_collected=56368123
2022-04-11T10:15:29.769286Z  INFO vector::sinks::blackhole::sink: Total events collected events=78402 raw_bytes_collected=65127690
2022-04-11T10:15:30.769155Z  INFO vector::sinks::blackhole::sink: Total events collected events=89104 raw_bytes_collected=74019078
2022-04-11T10:15:31.769552Z  INFO vector::sinks::blackhole::sink: Total events collected events=100308 raw_bytes_collected=83327888
2022-04-11T10:15:32.769753Z  INFO vector::sinks::blackhole::sink: Total events collected events=111493 raw_bytes_collected=92618588
```

Performance for the `vrl` condition type seems to be quite te same as in `0.20.0`, so the regression is still there.
However, I noticed that for the `check_fields` condition type performance seems better than even `0.19.3`.

I went to re-run only this type of condition again but **several times** for `0.19.3`, `0.20.0` and `0.20.1` each and I can confirm that actually this condition type is not affected by the regression described in the original issue. Apologies for reporting before that there was also a drop in performance with `check_fields`.

In summary, the regression seems to affect only `vrl` condition types.

**Comment by debugmiller** [comment]
@jszwedko anything I can do to help here? We'd like to upgrade to 0.20 due to https://github.com/vectordotdev/vector/issues/11871 but this is holding us back.

**Comment by jszwedko** [comment]
Thanks so much for providing the reproduction case @hhromic !

I do see a substantial regression between 0.19.3 and 0.20.1 for this case. 0.21.1 seems to make up a significant portion of the losses though, it's 15% slower rather than more than 100% slower. Capping the `demo_logs` at 10000 (`count`) and running with `hyperfine` I see:

```
0.19.3: 703ms +/- 6.9ms
0.20.1: 1524ms +/- 13ms
0.21.1: 803ms +/- 10ms
```

I'm going to dig in some more to see where the 0.19.3 -> 0.20.1 regression came from, since we might be able to make more of that back, but maybe 0.21.1's performance will be good enough for you @debugmiller ?

**Comment by jszwedko** [comment]
Also, apologies for the delay in response 😄

**Comment by jszwedko** [comment]
Ah, yes, I forgot that the `route` transform was refactored for 0.20.0 to, rather than run as one task per route, run as a single task that just iterates over all routes sequentially. I can see why this would regress `route` transforms that have a large number of routes. The `route` transform itself will still be parallelized but I'd hazard a guess that if the number of routes is > the number of `VECTOR_THREADS`, usually the number of available CPUs, then it is liable to perform worse.

I think we can actually get some of the performance back by running the condition checks concurrently. I opened https://github.com/vectordotdev/vector/issues/12481 to track this. Let me know if the performance of 0.21.1 is sufficient for you for now.

**Comment by jszwedko** [comment]
Running with just 8 routes (matching my thread count) I see:

0.19.3: 101.1 ms ±   2.2 ms
0.20.1: 156.9 ms ±   4.9 ms

With 4 routes:

0.19.3:  89.4 ms ±   3.4 ms
0.20.1: 107.7 ms ±   8.2 ms

Which seems to support my hypothesis that it the difference is worse the higher number of routes you have.

**Comment by debugmiller** [comment]
Thanks for the response! I’m taking a trip this weekend but I can test 0.21.1 when I get back and let you know.

**Comment by hhromic** [comment]
> Thanks so much for providing the reproduction case @hhromic !
>
Happy to help!
 
> I do see a substantial regression between 0.19.3 and 0.20.1 for this case. 0.21.1 seems to make up a significant portion of the losses though, it's 15% slower rather than more than 100% slower. Capping the `demo_logs` at 10000 (`count`) and running with `hyperfine` I see:

Didn't know about `hyperfine`. Just to confirm, is [this](https://github.com/sharkdp/hyperfine) its repository?
Looks like a really nice tool for any kind of benchmarking. Appreciate the tip!

I have one follow-up question to your comments that I will post in the new issue.

**Comment by jszwedko** [comment]
> Didn't know about `hyperfine`. Just to confirm, is [this](https://github.com/sharkdp/hyperfine) its repository?
> Looks like a really nice tool for any kind of benchmarking. Appreciate the tip!

That's the one!

**Comment by debugmiller** [comment]
Ok so *finally* I was able to test `0.21.1`. Performance is definitely better (I saw message throughput per core double), but I do still see my CPU utilization maxxing out at around 3.5-3.75 cores. To be honest this is good enough for me so I just shrunk my deployments down from `-t8` to `-t4`. 

I will definitely try out the next release with https://github.com/vectordotdev/vector/pull/12494. Seems like it dramatically improved performance!
